src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs

```diff
@@ -2205,19 +2205,22 @@ internal Node RemoveAll(Predicate<T> match)
                 Contract.Ensures(Contract.Result<Node>() != null);
 
                 var result = this;
-                int index = 0;
-                foreach (var item in this)
-                {
+                var index = 0;
+                var count = this.Count;
+                var removed = 0;
+                for(;;)
+                {
+                    if (index + removed >= count)
+                        break;
+                    var item = result[index];
```

Comment:
:memo: Indexing is O(log n), where `MoveNext()` on the enumerator is O(1), hence the noted change in complexity.